File: kotlin/RegretBasedLocalSearchModelEvaluation.kt
package hu.raven.puppet.job.experiments

import hu.raven.puppet.job.loadRegrets
import hu.raven.puppet.logic.step.bruteforce_solver.edgeBasedBranchAndBounds
import java.io.File
import kotlin.time.measureTimedValue


fun main() {
    (0..0).forEach { modelId ->

        val regretData = loadRegrets(
            File(
                "H:\\by-domain\\work_and_learning\\PhD\\research\\datasets",
                "2024_10_02_output_ATSP_trained_50_size_samples_100\\output_ATSP_trained_50_size_samples_100_size_250"
            )
        )

        val costDataPerNoiseLevelAndInstance =
            regretData
                .take(100)
                .mapIndexed { regretRecordIndex, regretRecord ->
                    measureTimedValue {
                        println("Regret record! $regretRecordIndex")
                        val (_, initialCost) = measureTimedValue {
                            edgeBasedBranchAndBounds(
                                regretRecord.distanceMatrix,
                                regretRecord.predictedRegretMatrix,//.transpose(),
                                true,
                            )
                        }.let {
                            println("Finding initial solution took: ${it.duration}")
                            it.value
                        }
                        val (permutation, cost) = measureTimedValue {
                            edgeBasedBranchAndBounds(
                                regretRecord.distanceMatrix,
                                regretRecord.predictedRegretMatrix,//.transpose(),
                                false,
                            )
                        }.let {
                            println("Refining with branch and bounds took: ${it.duration}")
                            it.value
                        }

                        println(
                            "COST ${
                                calcCostOfTspSolution(
                                    permutation,
                                    regretRecord.distanceMatrix
                                )
                            }"
                        )

                        var bestCost = cost

                        repeat(1) {
                            //println(it)
                            bestCost = threeOptCycle(
                                permutation,
                                regretRecord,
                                regretRecord.predictedRegretMatrix.transpose(),
                                cost
                            )
                        }

                        CostRecord(
                            initialCost = initialCost,
                            builtCost = cost,
                            optimizedCost = bestCost,
                            optimal = regretRecord.optCost
                        )
                    }.also {
                        println("processing instance took: ${it.duration}")
                    }
                }
                .toList()
        println(
            costDataPerNoiseLevelAndInstance.map { it.value.optimizedCost / it.value.optimal - 1.0 }.toDoubleArray().statistics()
        )
        println(
            costDataPerNoiseLevelAndInstance.map { it.duration.inWholeMilliseconds.toDouble() }.toDoubleArray().statistics()
        )
    }
}

File: src/regret_based_local_search_model_evaluation.cpp
#include <iostream>

#include "calc_cost_of_tsp.h"
#include "cost_record.h"
#include "edge_based_branch_and_bounds.h"
#include "permutation.h"
#include "regret_data.h"
#include "three_opt_cycle.h"

int main() {
    for (int model_index = 0; model_index <= 1; model_index++) {
        auto regretData = loadRegrets(
            "H:\\by-domain\\work_and_learning\\PhD\\research\\datasets\\2024_10_02_output_ATSP_trained_50_size_samples_100\\output_ATSP_trained_50_size_samples_100_size_250"
        );

        vector<cost_record> results =
                views::zip(views::iota(0, static_cast<int>(regretData.size())), regretData)
                | views::take(100)
                | views::transform([](std::pair<int, regret_data> item) {
                    auto [regretRecordIndex, regretRecord] = item;
                    cout << "Regret record! " << regretRecordIndex << endl;
                    auto [_,initial_cost] = edge_based_branch_and_bounds(
                        regretRecord.getDistanceMatrix(),
                        regretRecord.getPredictedRegretMatrix(),
                        true
                    );
                    auto [solution,cost] = edge_based_branch_and_bounds(
                        regretRecord.getDistanceMatrix(),
                        regretRecord.getPredictedRegretMatrix(),
                        false
                    );
                    cout << "COST: " << calc_cost_of_tsp(
                        solution, regretRecord.getDistanceMatrix()) << endl;
                    auto best_cost = cost;

                    for (auto repeat = 0; repeat < 1; repeat++) {
                        best_cost = three_opt_cycle(
                            solution,
                            regretRecord,
                            regretRecord.getPredictedRegretMatrix(),
                            cost
                        );
                    }

                    return cost_record(initial_cost, cost, best_cost, regretRecord.getOptCost());
                })
                | ranges::to<vector<cost_record> >();

        cout << "initial,pre_optimized,optimized,optimal" << endl;
        for (auto result: results) {
            cout << result.getInitialCost() <<
                    "," << result.getBuiltCost() <<
                    "," << result.getOptimizedCost() <<
                    "," << result.getOptimal() <<
                    endl;
        }
    }
}
